### PR TITLE
update crucible-ci to use "massively parallel" model

### PR DESCRIPTION
**Pull request by k-rister**

### CODE REVIEW

File: .github/workflows/crucible-ci.yaml [hunk 1]
```diff
@@ -9,29 +9,83 @@ on:
...
+    - generate-job-matrix-parameters
+    - display-job-matrix-parameters
+    continue-on-error: true
+    strategy:
+      matrix:
+        endpoint: ${{ fromJSON(needs.generate-job-matrix-parameters.outputs.endpoints_self) }}
+        userenv: ${{ fromJSON(needs.generate-job-matrix-parameters.outputs.userenvs) }}
     steps:
-    - name: Checkout bench-oslat
+    - name: Matrix Parameters => (${{ matrix.endpoint }}, ${{ matrix.userenv }})
```
Comment (rafaelfolco) [inline .github/workflows/crucible-ci.yaml:59]: nit: matrix (lower case?)

File: .github/workflows/crucible-ci.yaml [hunk 2]
```diff
@@ -9,29 +9,83 @@ on:
...
+  crucible-ci-self-hosted-runners:
+    runs-on: [ self-hosted ]
+    needs:
+    - generate-job-matrix-parameters
+    - display-job-matrix-parameters
+    continue-on-error: true
+    strategy:
+      matrix:
+        endpoint: ${{ fromJSON(needs.generate-job-matrix-parameters.outputs.endpoints_self) }}
+        userenv: ${{ fromJSON(needs.generate-job-matrix-parameters.outputs.userenvs) }}
```
Comment (rafaelfolco) [inline .github/workflows/crucible-ci.yaml:57]: does this mean these steps will be executed in parallel for all endpoints ?

File: .github/workflows/crucible-ci.yaml [hunk 3]
```diff
@@ -9,29 +9,83 @@ on:
   workflow_dispatch:
 
 jobs:
-  crucible-ci:
-    runs-on: [self-hosted, cpu-partitioning, remotehost]
+  generate-job-matrix-parameters:
+    runs-on: ubuntu-latest
+    outputs:
+      userenvs: ${{ steps.get-userenvs.outputs.userenvs }}
```
Comment (rafaelfolco) [inline .github/workflows/crucible-ci.yaml:15]: steps.get-userenvs would be more intuitive... steps.get-userenvs.outputs.userenvs is weird

**Review by rafaelfolco — APPROVED** [review]

File: .github/workflows/crucible-ci.yaml (hunk 1 above)
Comment (k-rister) [reply .github/workflows/crucible-ci.yaml:59]: not sure I follow what you mean here...

File: .github/workflows/crucible-ci.yaml (hunk 2 above)
Comment (k-rister) [reply .github/workflows/crucible-ci.yaml:57]: It is kinda like multiplex, it takes the arrays defined under a strategy (named 'matrix') and creates a job for all possible combinations.  The jobs are all queued up at the same time (somewhat, it takes a bit longer to create them than you might think, on the order of 10s of seconds).  How many actually run simultaneously is governed by the Github runner limits, for our `perftool-incubator` project that is 20.

File: .github/workflows/crucible-ci.yaml (hunk 3 above)
Comment (k-rister) [reply .github/workflows/crucible-ci.yaml:15]: I hear you, but this is the way it is.  Each `step` can have one or more outputs defined so the are "collected" under a "container"/"object" called "outputs" (my poor attempt at describing how it works).

File: .github/workflows/crucible-ci.yaml (hunk 1 above)
Comment (rafaelfolco) [reply .github/workflows/crucible-ci.yaml:59]: its just a nit picking... follow the same pattern: if you renamed step name to lower case in the other cases, why not here?

Comment (k-rister) [reply .github/workflows/crucible-ci.yaml:59]: fair enough... ;-)

Comment (rafaelfolco) [reply .github/workflows/crucible-ci.yaml:59]: **m**atrix parameters =>

**Pull request merged**